README.md
# mongo-monitor

[![CircleCI](https://circleci.com/gh/dwmkerr/mongo-monitor.svg?style=shield)](https://circleci.com/gh/dwmkerr/mongo-monitor) [![codecov](https://codecov.io/gh/dwmkerr/mongo-monitor/branch/master/graph/badge.svg)](https://codecov.io/gh/dwmkerr/mongo-monitor) [![Greenkeeper badge](https://badges.greenkeeper.io/dwmkerr/mongo-monitor.svg)](https://greenkeeper.io/) [![GuardRails badge](https://badges.guardrails.io/dwmkerr/mongo-monitor.svg?token=569f2cc38a148f785f3a38ef0bcf5f5964995d7ca625abfad9956b14bd06ad96&provider=github)](https://dashboard.guardrails.io/default/gh/dwmkerr/mongo-monitor)

Simple CLI to monitor the status of a MongoDB cluster.

```bash
mongo-monitor mongodb://address:27017
```

![Replicaset Screenshot](./docs/overview.gif)

The connection string is handled with [`mongo-connection-string`](https://github.com/dwmkerr/mongo-connection-string), which means it'll handle input which is not URI encoded.

Install with:

```bash
npm install mongo-monitor
```

## Usage

The monitor is primary designed to show the status of a MongoDB cluster, updated real-time. This is useful when performing administrative operations such as replicaset or shard configuration.

On a sharded cluster, if you provide a connection string with admin priviledges to any `mongos` host, you will see the sharding configuration.

On a replicaset, if you provide a connection string with admin priviledges to any host, or to the entire set, you will see the replicaset configuration:

![Replicaset Screenshot](./docs/replicaset.jpg)

For a standalone, basic info is reported. Sharded clusters show details of the shards, as welll as the replicasets which make up each shard:

![Sharded Cluster Screenshot](./docs/sharded-cluster.jpg)

## Samples

To try the monitor out, there are a few samples you can try.

Monitor a standalone instance:

```bash
make sample-standalone
mongo-monitor localhost:271017
```

Monitor a replicaset:

```bash
make sample-replicaset
mongo-monitor localhost:27017,localhost:27018,localhost:27019?replicaSet=cluster
```

Monitor a sharded cluster:

```bash
make sample-shard
mongo-monitor localhost
```

Cleanup sample processes:

```bash
make sample-shutdown
```

## Developing

Lint with `npm run lint`, test with `npm test`, release with `npm run release`.

## Tests

The following files are useful for testing:

| File | Notes |
|------|-------|
| `shard.isMaster.json` | The output of `isMaster` for a `mongos` member of a sharded cluster. | 
| `shard.node.isMaster.json` | The output of `isMaster` for a `mongod` member of a sharded cluster. | 
| `shard.listShards.json` | The output of `listShards` for a `mongos` member of a sharded cluster. | 
| `shard.replset.replSetGetStatus.json` | The output of `replSetGetStatus` for a replicaset in a sharded cluster. |
| `replset.isMaster.json` | The output of `isMaster` for a replicaset. |
| `replset.replSetGetStatus.json` | The output of `replSetGetStatus` for a replicaset. |
| `replset.serverStatus.json` | The output of `serverStatus` for a replicaset. |
| `standalone.isMaster.json` | The output of `isMaster` for a standalone server. |
| `standalone.serverStatus.json` | The output of `serverStatus` for a standalone server. |

## Troubleshooting

```bash
mongo-monitor : command not found
```

You need to change the npm config prefix

```bash
npm config set prefix /usr/local

npm root -g
```


## Notes

There's also a little blog post on this here: [dwmkerr.com/mongo-monitor-cli/](https://www.dwmkerr.com/mongo-monitor-cli/).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.1](https://github.com/dwmkerr/mongo-monitor/compare/v1.1.0...v1.1.1) (2019-11-05)

## [1.1.0](https://github.com/dwmkerr/mongo-monitor/compare/v1.0.0...v1.1.0) (2019-11-05)


### Features

* show connection statistics ([279fc43](https://github.com/dwmkerr/mongo-monitor/commit/279fc4397a872a001ae3e15a6ae6cea354c7fd58))

<a name="1.0.0"></a>
# [1.0.0](https://github.com/dwmkerr/mongo-monitor/compare/v0.3.0...v1.0.0) (2018-06-23)



<a name="0.3.0"></a>
# [0.3.0](https://github.com/dwmkerr/mongo-monitor/compare/v0.2.1...v0.3.0) (2018-05-15)


### Bug Fixes

* **sharded-cluster:** correctly show shard replset status ([17c2a1f](https://github.com/dwmkerr/mongo-monitor/commit/17c2a1f)), closes [#8](https://github.com/dwmkerr/mongo-monitor/issues/8)


### Features

* support standalone shards ([cb3e058](https://github.com/dwmkerr/mongo-monitor/commit/cb3e058))
* **event-logging:** log topology events, cleanup structure ([c15b649](https://github.com/dwmkerr/mongo-monitor/commit/c15b649))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/dwmkerr/mongo-monitor/compare/v0.2.0...v0.2.1) (2018-05-02)


### Bug Fixes

* **connection-string:** use the `mongodb://` protocol by default ([886b05f](https://github.com/dwmkerr/mongo-monitor/commit/886b05f))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/dwmkerr/mongo-monitor/compare/v0.1.0...v0.2.0) (2018-03-14)


### Features

* improve connection string handling ([68f8d41](https://github.com/dwmkerr/mongo-monitor/commit/68f8d41))



<a name="0.1.0"></a>
# 0.1.0 (2018-03-10)


samples/replicaset/replicaset.sh
#!/usr/bin/env bash

# Basic config.
port1=27017
port2=27018
port3=27019
port4=27020

# Rebuild data and log directories.
if [[ -d './data' ]]; then
    rm -rf ./data
fi
if [[ -d './log' ]]; then
    rm -rf ./log
fi
mkdir -p ./data/data1
mkdir -p ./data/data2
mkdir -p ./data/data3
mkdir -p ./data/data4
mkdir -p ./log

# Start the replicaset processes.
mongod --fork --logpath ./log/node1.log --oplogSize 50 --port ${port1} --dbpath ./data/data1 --replSet cluster
mongod --fork --logpath ./log/node2.log --oplogSize 50 --port ${port2} --dbpath ./data/data2 --replSet cluster
mongod --fork --logpath ./log/node3.log --oplogSize 50 --port ${port3} --dbpath ./data/data3 --replSet cluster
mongod --fork --logpath ./log/node4.log --oplogSize 50 --port ${port4} --dbpath ./data/data4 --replSet cluster

# Wait a bit of time for the process to start.
sleep 3

# Note that we don't add node 4, so we can easily demo adding a node.
mongo --port 27017 --shell <<- EOF
rs.initiate({
  _id: 'cluster',
  members: [{
    _id: 0,
    host: 'localhost:27017'
  },{
    _id: 1,
    host: 'localhost:27018'
  },{
    _id: 2,
    host: 'localhost:27019',
    arbiterOnly: true
  }]
});
EOF

# Show the process info.
echo
echo "Replicaset MongoDB Running"
echo
echo "Monitor with: "
echo "  mongo-monitor localhost:${port1},localhost:${port2},localhost:${port3}?replicaSet=cluster"

